daos: dcp: enhanced error checking

daltonbohning opened the pull request:
My branch deviated a bit from my previous PR, I had many small commits, and git wasn't cooperating, so I'm just creating a new PR entirely.

**common/mfu_errors.h**
- New file
- Defines some generic error codes, codes for dcp, and codes for DAOS
  - Ranges are in the 1000s, 1100s, and 4000s
- Defines MFU_ERRF and MFU_ERRP for printing the error codes.
  - Formatted as: MFU_ERR(#)
  - Could be improved in the future to print the"string" error code instead of just MFU_ERR
    - DAOS has something similar to this

**common/mfu_io**
- Defined daos_access

**common/mfu_param_path.c**
- mfu_param_path_check_copy - Return on DAOS error

**common/mfu_util**
- daos_connect
  - Added flag for whether or not to connect to pool
  - Added flag for whether or not to auto-create the destination container
    - No longer creates the source container automatically
  - These flags allow a centralized way to capture errors. I.e. only one bcast is needed for pool connection, cont connection, cont creation
  - Rank 0 now tracks and broadcasts any errors before trying to share the handles
  - Function returns an error

**dcp/dcp.c**
- Use the new error codes from mfu_error.h
- DAOS errors are generally handled more precisely
  - Each process tracks local errors, and stops making DAOS calls once it has run into an error
  - Allreduce on local DAOS errors, so all processes will know to exit
  - Check DAOS arguments before trying to make DAOS calls
  - Refactored daos_connect calls to better capture errors and exit if found
  - More gracefully handle a failed daos_init call
  - Exit if DAOS source is DAOS destination

closes #363 
closes #361
closes #370

adammoody commented on src/common/mfu_io.c:
```diff
@@ -227,11 +227,23 @@ int mfu_access(const char* path, int amode)
+    parse_filename(path, &name, &dir_name);
+    assert(dir_name);
+
+    dfs_obj_t *parent = NULL;
```
```suggestion
    dfs_obj_t* parent = NULL;
```

adammoody commented on src/common/mfu_util.c:
```diff
@@ -107,58 +107,89 @@ void daos_bcast_handle(
+                    poh, &pool_info, NULL);
+            if (rc != 0) {
+                MFU_LOG(MFU_LOG_ERR, "Failed to connect to pool");
+                goto bcast;
```
do we want a d_rank_list_free(svcl) call in here?

daltonbohning replied:
Yes, good catch!

adammoody commented on src/common/mfu_util.c:
```diff
@@ -107,58 +107,89 @@ void daos_bcast_handle(
+            if (rc != 0) {
+                MFU_LOG(MFU_LOG_ERR, "Failed to connect to pool");
+                goto bcast;
+            }
+            d_rank_list_free(svcl);
         }
-        d_rank_list_free(svcl);
 
-        /* attempt to open the daos container */
+        /* Try to open the container */
         daos_cont_info_t co_info;
         rc = daos_cont_open(*poh, cont_uuid, DAOS_COO_RW, coh, &co_info, NULL);
-
+        
         /* If NOEXIST we create it */
```
Let's pull the lines that set ``rc =`` and the corresponding ``if (rc != 0)`` statement that checks its value right beside each other without any empty lines in between.  I find that helps with long term code maintenance in avoiding someone from inserting code in between that might reuse rc before it is checked.

daltonbohning replied:
Ah, yes, I agree!

adammoody commented on src/common/mfu_util.c:
```diff
@@ -107,58 +107,89 @@ void daos_bcast_handle(
+
+    /* return if invalid */
+    if (valid == 0) {
+        return -1;
```
The valid check to guard the handle bast looks good.  Thanks.  One other thing to check is whether we should clean up any resources that may have been allocated if we fail.  I don't know the daos api well enough to check, but I did see one call that looked to trying to free something above which would not be called in certain error cases.

daltonbohning replied:
Yes, you are right. I missed that. Thank you!

daltonbohning replied:
Luckily, I think only the d_rank_list_free was missing here, since the pool/container handles are declared statically as structs in dcp.c, and the daos calls expect them to not be NULL, so I don't believe the daos calls allocate anything else here.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -22,7 +22,120 @@
+            if (rank == 0) {
+                MFU_LOG(MFU_LOG_ERR, "Using same pool uuid with different svcl's");
+            }
+        rc = 1;
```
```suggestion
            rc = 1;
```

adammoody commented on src/dcp/dcp.c:
```diff
@@ -22,7 +22,120 @@
+            if (rank == 0) {
+                 MFU_LOG(MFU_LOG_ERR, "DAOS source is DAOS destination");
+            }
+        rc = 1;
```
```suggestion
            rc = 1;
```

adammoody commented on src/dcp/dcp.c:
```diff
@@ -470,104 +595,160 @@ int main(int argc, char** argv)
-    if (mfu_src_file->type == DAOS) {
-        /* if DAOS source pool uuid is valid, then set source file type to DAOS */
-        daos_connect(rank, src_svc, src_pool_uuid, src_cont_uuid, &src_poh, &src_coh); 
+    if (!local_daos_error && mfu_src_file->type == DAOS) {
```
Skimming the code, I get the impression that local_daos_error is tracking whether the calling process has seen an error, which may be independent of other processes.  Is that right?

daltonbohning replied:
Correct. It is *technically possible* that some DAOS call might fail for only a single process, some of which might stem from erroneous errors. Practically speaking, most processes at the point of the broadcast into global_daos_error are *most likely* in a similar state, but to be safe in calling MPI_Finalize(), etc., it's best to make sure. 
One such out-of sync issue that could possibly arise, though unlikely, would be in our call to duns_resolve_path. 
Let's say we have 2 processes. One of them calls duns_resolve_path and succeeds. Before the other process makes the call, the UNS path is removed from the DAOS system, and then the 2nd process makes the call, and fails. Now the processes are in different states.
This is where my reasoning for "local_daos_error" comes from, among other erroneous network errors that might crop up.

adammoody replied:
Assuming that's the case, one thing to be careful of is mixed error conditions, like if local_daos_error=true on rank=0, but false on rank=1.  It looks like daos_connect includes a collective (bcast), but then you might have rank 0 skipping that call while rank 1 does.

adammoody replied:
> 
> 
> Correct. It is _technically possible_ that some DAOS call might fail for only a single process, some of which might stem from erroneous errors. Practically speaking, most processes at the point of the broadcast into global_daos_error are _most likely_ in a similar state, but to be safe in calling MPI_Finalize(), etc., it's best to make sure.
> One such out-of sync issue that could possibly arise, though unlikely, would be in our call to duns_resolve_path.
> Let's say we have 2 processes. One of them calls duns_resolve_path and succeeds. Before the other process makes the call, the UNS path is removed from the DAOS system, and then the 2nd process makes the call, and fails. Now the processes are in different states.
> Another such issue might be with mounting, since we could possibly run out of mountpoints for only one process.
> This is where my reasoning for "local_daos_error" comes from, among other erroneous network errors that might crop up.

Makes sense.  Good thinking.

daltonbohning replied:
Oh yes... I'll need to rework that portion.

adammoody replied:
This kind of check comes up a lot.  This utility function comes in handy:
https://github.com/hpc/mpifileutils/blob/a9cd09d6a4412b5a5e87419c8423d6d79f32caab/src/common/mfu_util.c#L1096

You can use it like `` if (! mfu_alltrue(fd >= 0, MPI_COMM_WORLD)) ``

daltonbohning replied:
This is the full function:
https://github.com/daltonbohning/mpifileutils/blob/6f4a6d3252b720c770ed2bcb64c99dac3ae9f2bf/src/common/mfu_util.c#L1125-L1133

Then, I would use it like `if ! (mfu_alltrue(!local_daos_error, MPI_COMM_WORLD))`

But this function uses `MPI_INT`, `MPI_LAND`, whereas mine was using `MPI_C_BOOL`, `MPI_LOR`.
Do you know if there would be any performance difference here? I.e. perhaps MPI has an optimized implementation for `MPI_C_BOOL`, `MPI_LOR` vs `MPI_INT`, `MPI_LAND`?

daltonbohning replied:
I actually found another hole in the logic here, where `same_pool` could potentially be mixed if a process has an error, and so the daos_connect call might be mixed. I ended up putting another reduction on `local_daos_error` before calling `daos_connect`. Even if refactored, I couldn't think of another way around this that was foolproof.

adammoody replied:
> 
> 
> This is the full function:
> https://github.com/daltonbohning/mpifileutils/blob/6f4a6d3252b720c770ed2bcb64c99dac3ae9f2bf/src/common/mfu_util.c#L1125-L1133
> 
> Then, I would use it like `if ! (mfu_alltrue(!local_daos_error, MPI_COMM_WORLD))`
> 
> But this function uses `MPI_INT`, `MPI_LAND`, whereas mine was using `MPI_C_BOOL`, `MPI_LOR`.
> Do you know if there would be any performance difference here? I.e. perhaps MPI has an optimized implementation for `MPI_C_BOOL`, `MPI_LOR` vs `MPI_INT`, `MPI_LAND`?
> Or even just a difference between only the datatypes? Or only the `LAND` vs `LOR`? _**Theoretically**_, `MPI_C_BOOL`, `MPI_LOR` _**could**_ be implemented in a more efficient manner.

Yes, that's right.  One could implement an LAND/LOR allreduce on a bool using a single bit versus the 32-bits an int type would require.  It should be fine to modify the mfu_alltrue function to take a bool as its input if you're up for it.

daltonbohning replied:
Sure, I'll do that (after a quick sanity check to everything calling mfu_alltrue).

daltonbohning commented:
Thanks for all the input @adammoody !
I should have these things all resolved tomorrow.

daltonbohning commented:
I made some style updates, corrected the "hole" with the way `local_daos_error` and `daos_connect` were being handled, and also corrected another hole with `local_daos_error` and `same_pool`.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -496,76 +711,68 @@ int main(int argc, char** argv)
+    if (!local_daos_error && mfu_dst_file->type == DAOS) {
+        if (same_pool) {
+            /* Don't reconnect to pool, but do create container if non-existent */
+            if (rank == 0)
+                rc = daos_connect(rank, dst_svc, dst_pool_uuid, dst_cont_uuid, &src_poh, &dst_coh, false, true);
+            else
+                rc = daos_connect(rank, dst_svc, dst_pool_uuid, dst_cont_uuid, &src_poh, &dst_coh, false, true);
```
Just another style thing is that we brace even one-line code blocks, again to avoid future bugs sneaking in where someone later adds a second line in the block without realizing the braces are missing.
```suggestion
            if (rank == 0) {
                rc = daos_connect(rank, dst_svc, dst_pool_uuid, dst_cont_uuid, &src_poh, &dst_coh, false, true);
            } else {
                rc = daos_connect(rank, dst_svc, dst_pool_uuid, dst_cont_uuid, &src_poh, &dst_coh, false, true);
            }
```

daltonbohning replied:
This one is interesting... That was a quick debug I was doing, but then removed it. There should only be a singel `daos_connect` call there, and my local branch has the correct version...
I'm investigating now to see why git ignored these changes...

daltonbohning replied:
Okay this was very strange. My local repo was "up to date", but still did not match the branch on GitHub.
Anyway, I fixed it. Thanks for the catch!

adammoody commented:
@daltonbohning , thanks for checking those other return code cases.  I made one more style request, but otherwise it looks good to me.

The DAOS code that leads into the tool is getting long enough that we can think about abstracting that into a function (starting at line 632 in dcp.c).  That can be done later, though.

daltonbohning commented:
@adammoody You're welcome, but really, thank you for catching this! Another set of eyes definitely helps.
I've had a similar thought about abstracting the daos-specific code. I'll go ahead and make an issue to keep track of this, and I'll make those changes when I get a chance, especially since we will be making more daos changes and additions in the future.

I pushed a couple more things in the last few minutes. If you don't find anything else, I don't have anything more to add to this PR.

adammoody commented:
No problem.  Those MPI corner cases can be tricky.

Thanks, @daltonbohning .  Looks good to me.  Would you mind squashing those commits into one and force pushing back?

adammoody commented:
Oh, you can leave the mfu_alltrue commit separate in the squash since that's a different fix.

daltonbohning commented:
Sorry, I'm working on this. My repo seemed to have a breakdown and it's taking a lot of time to squash these commits.
(It explains the strange out-of-sync issue I ran into earlier)
